File: ontrack-extension-issues/src/main/java/net/nemerosa/ontrack/extension/issues/model/ConfiguredIssueService.java
package net.nemerosa.ontrack.extension.issues.model;

import lombok.Data;
import net.nemerosa.ontrack.extension.issues.IssueServiceExtension;
import net.nemerosa.ontrack.model.structure.Project;
import net.nemerosa.ontrack.model.support.MessageAnnotationUtils;
import net.nemerosa.ontrack.model.support.MessageAnnotator;

import java.util.*;

/**
 * Association between an {@link net.nemerosa.ontrack.extension.issues.IssueServiceExtension} and
 * one of its {@link net.nemerosa.ontrack.extension.issues.model.IssueServiceConfiguration configuration}s.
 */
@Data
public class ConfiguredIssueService {

    private final IssueServiceExtension issueServiceExtension;
    private final IssueServiceConfiguration issueServiceConfiguration;

    public String formatIssuesInMessage(String message) {
        return issueServiceExtension.getMessageAnnotator(issueServiceConfiguration)
                .map(annotator -> MessageAnnotationUtils.annotate(message, Collections.singletonList(annotator)))
                .orElse("");
    }

    public String getLinkForAllIssues(List<Issue> issues) {
        return issueServiceExtension.getLinkForAllIssues(issueServiceConfiguration, issues);
    }

    public Issue getIssue(String issueKey) {
        return issueServiceExtension.getIssue(issueServiceConfiguration, issueKey);
    }

    public IssueServiceConfigurationRepresentation getIssueServiceConfigurationRepresentation() {
        return IssueServiceConfigurationRepresentation.Companion.of(
                issueServiceExtension,
                issueServiceConfiguration
        );
    }

    public Optional<MessageAnnotator> getMessageAnnotator() {
        return issueServiceExtension.getMessageAnnotator(issueServiceConfiguration);
    }

    /**
     * @deprecated Do not use
     */
    @Deprecated
    public Set<String> extractIssueKeysFromMessage(String message) {
        return issueServiceExtension.extractIssueKeysFromMessage(issueServiceConfiguration, message);
    }

    /**
     * @deprecated Do not use
     */
    @Deprecated
    public boolean containsIssueKey(String key, Set<String> keys) {
        return issueServiceExtension.containsIssueKey(issueServiceConfiguration, key, keys);
    }

    public Optional<String> getIssueId(String token) {
        return issueServiceExtension.getIssueId(issueServiceConfiguration, token);
    }

    public Collection<? extends Issue> getLinkedIssues(Project project, Issue issue) {
        return issueServiceExtension.getLinkedIssues(project, issueServiceConfiguration, issue);
    }
}

File: ontrack-extension-issues/src/main/java/net/nemerosa/ontrack/extension/issues/model/ConfiguredIssueService.kt
package net.nemerosa.ontrack.extension.issues.model

import net.nemerosa.ontrack.extension.issues.IssueServiceExtension
import net.nemerosa.ontrack.extension.issues.model.IssueServiceConfigurationRepresentation.Companion.of
import net.nemerosa.ontrack.model.support.MessageAnnotator

/**
 * Association between an [net.nemerosa.ontrack.extension.issues.IssueServiceExtension] and
 * one of its [configuration][net.nemerosa.ontrack.extension.issues.model.IssueServiceConfiguration]s.
 */
data class ConfiguredIssueService(
    val issueServiceExtension: IssueServiceExtension,
    val issueServiceConfiguration: IssueServiceConfiguration,
) {

    fun getIssue(issueKey: String): Issue? {
        return issueServiceExtension.getIssue(issueServiceConfiguration, issueKey)
    }

    val issueServiceConfigurationRepresentation: IssueServiceConfigurationRepresentation
        get() = of(
            issueServiceExtension,
            issueServiceConfiguration
        )

    val messageAnnotator: MessageAnnotator?
        get() = issueServiceExtension.getMessageAnnotator(issueServiceConfiguration)

    fun extractIssueKeysFromMessage(message: String): Set<String> {
        return issueServiceExtension.extractIssueKeysFromMessage(issueServiceConfiguration, message)
    }

    /**
     * Given an issue key, returns its display form.
     *
     * @param key Key ID
     * @return Display key
     */
    fun getDisplayKey(key: String): String {
        return issueServiceExtension.getDisplayKey(issueServiceConfiguration, key)
    }

}
